solutions/USACO dec16-lasers-mirrors.mdx

```diff
@@ -0,0 +1,79 @@
+---
+id: usaco-671
+source: USACO Gold 2016 December
+title: Lasers and Mirrors
+author: Maggie Liu
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_lasers_gold_dec16.html)
+
+It is optimal to use each fence post at most once, so we can use BFS to find the minimum number of fence posts needed to direct the laser to the barn. We first map a vector of fence post indices to their x and y coordinates, which we store in $\texttt{lines}$. In the queue $q$, we store the index and a boolean representing the direction of the incoming beam for each point, with true representing a horizonal incoming beam and false representing a vertical incoming beam. The array $\texttt{dist}$ will store the number of edges of the shortest path from the laser to each point.
+
+When processing an element in the queue, we determine the direction of the incoming beam and iterate the points that the beam can be deflected to. For each of these points, if it is unvisited, we add it into the queue and update the distance as one more than the current distance.
+
+If the distance to the barn, $\texttt{dist[1]}$, is $10^9$, we aren't able to reach the barn so we print $-1$. If we are able to, the number of mirrors needed is one less than the distance.
```

Comment:
I don't use INF in my code, so it might be unclear?


I think the first one looks better?
![image](https://user-images.githubusercontent.com/63619830/110030035-0f8ca880-7cea-11eb-8c28-f7197dfd81e0.png)

Reply:
> I don't use INF in my code, so it might be unclear?

okay

> I think the first one looks better?

yeah, ig $\texttt{dist[1]}$ is good